### PR TITLE
init generated column grammar and other checks.

### PR DESCRIPTION
**Pull request by hicqu**
ddl implements:
- [x] create table t(a int generated as (b+1))    `-- if b is not a valid field, report error.`
- [x] alter table t add b int as (a+1);                  `-- if a is not a valid field, report error.`
- [x] alter table t drop column a;                       `-- report error if a is dependent by other fields.`
- [x] create table t(a json, key (a))   `-- report error because cannot index json field`
- [x] update talbe t set generated_field = 3; `--report error when update generated field`
- [x] insert into t (json_field) values (3); -- `report error when insert a generated field`
- [ ] select ... from t where expression; `-- if the expression is equal to some generated columns, use index`
- [ ] JSON cannot has default value in DDL;

### CODE REVIEW

**Comment by hicqu** [comment]
initialize generated columns. PTAL, thanks.

File: plan/resolver.go [hunk 1]
```diff
@@ -75,6 +76,8 @@ type resolverContext struct {
 	fieldList []*ast.ResultField
 	// result fields collected in group by clause.
 	groupBy []*ast.ResultField
+	// column defines collected in create/alter statement.
```
Comment (XuHuaiyu) [inline plan/resolver.go]: s/defines/definitions ?

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -437,6 +443,76 @@ func checkDuplicateColumn(colDefs []*ast.ColumnDef) error {
...
+// If colDef depends on itself, it will return an error.
+func findDependedColumnNames(colDef ast.ColumnDef) (cols []string, err error) {
+	var colsMap = make(map[string]bool, 0)
+	for _, option := range colDef.Options {
+		if option.Tp == ast.ColumnOptionGenerated {
+			colNames := findColumnNamesInExpr(option.Expr)
+			for _, depCol := range colNames {
+				if colDef.Name.Name.L == depCol.Name.L {
+					err = errBadField.GenByArgs(depCol.Name.L, "generated column function")
+					break
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: return?

File: parser/parser.y [hunk 3]
```diff
@@ -114,6 +114,7 @@ import (
 	foreign			"FOREIGN"
 	from			"FROM"
 	fulltext		"FULLTEXT"
+	generated		"GENERATED"
```
Comment (tiancaiamao) [inline parser/parser.y]: Reserved Keyword

File: parser/parser.y [hunk 4]
```diff
@@ -197,6 +198,7 @@ import (
 	smallIntType		"SMALLINT"
 	starting		"STARTING"
 	tableKwd		"TABLE"
+	stored			"STORED"
```
Comment (tiancaiamao) [inline parser/parser.y]: Reserved Keyword

File: parser/parser.y [hunk 5]
```diff
@@ -217,6 +219,7 @@ import (
 	values			"VALUES"
 	varcharType		"VARCHAR"
 	varbinaryType		"VARBINARY"
+	virtual			"VIRTUAL"
```
Comment (tiancaiamao) [inline parser/parser.y]: Reserved Keyword

File: parser/parser.y [hunk 6]
```diff
@@ -232,6 +235,7 @@ import (
 	addTime				"ADDTIME"
 	admin				"ADMIN"
 	aesDecrypt			"AES_DECRYPT"
+	always				"ALWAYS"
```
Comment (tiancaiamao) [inline parser/parser.y]: NonReserved Keyword

Comment (tiancaiamao) [inline parser/parser.y]: You need add them to 
`UnReservedKeyword:`
and 
`ReservedKeyword:` 
in parser.y

And also add them to `TestSimple` in parser_test.go

If you're not sure whether a ident is keyword or not, look up it here:
https://dev.mysql.com/doc/refman/5.7/en/keywords.html

File: parser/parser.y [hunk 7]
```diff
@@ -1293,6 +1298,23 @@ ColumnOption:
...
+		$$ = &ast.ColumnOption{
+			Tp: ast.ColumnOptionGenerated,
+			Expr: expr,
+			Stored: $6.(bool),
+		}
+	}
+
+GeneratedAlways: | "GENERATED" "ALWAYS"
+
+VirtualOrStored: { $$ = false }| "VIRTUAL" { $$ = false }| "STORED" { $$ = true }
```
Comment (tiancaiamao) [inline parser/parser.y]: Split into multiple lines.

File: util/stringutil/string_util.go [hunk 8]
```diff
@@ -239,3 +240,10 @@ func DoMatch(str string, patChars, patTypes []byte) bool {
 	}
 	return sIdx == len(str)
 }
+
+var reBlank = regexp.MustCompile(`\s`)
+
+// RemoveBlanks remove all blanks, returns a new string.
+func RemoveBlanks(s string) string {
```
Comment (tiancaiamao) [inline util/stringutil/string_util.go]: Don't forget unit test when you add something.
Personally speaking, I dislike regexp for it's poor performance.

**Comment by hicqu** [comment]
should split this PR into several small things.

**Pull request closed without merging**